include/dt-bindings/clock/microchip,mpfs-clock.h

```diff
@@ -54,18 +54,19 @@
 
 #define CLK_CCC_PLL0		0
 #define CLK_CCC_PLL1		1
-#define CLK_CCC_DLL0		2
-#define CLK_CCC_DLL1		3
 
-#define CLK_CCC_PLL0_OUT0	4
-#define CLK_CCC_PLL0_OUT1	5
-#define CLK_CCC_PLL0_OUT2	6
-#define CLK_CCC_PLL0_OUT3	7
+#define CLK_CCC_PLL0_OUT0	2
```

Comment:
Without looking too closely, I figure something like changing mpfs_ccc_register_outputs() so that it doesn't add to the hws array using out_hw->id, but rather a shifted down version of out_hw->id would work?

Reply:
> Unfortunately, these defines are an ABI and therefore cannot be changed. The bug needs to be solved without changing the binding header, only by changing the driver.

Which part is dependent on this?

I guess it has to allocate `num_clks + 2 (where num_clks=10)` to store `hw_data.hws[10]` and `hw_data.hws[11]` for `CLK_CCC_PLL1_OUT2/3`.
`hw_data.hws[2]` and `hw_data.hws[3]` will be NULL.
```c
clk_data = devm_kzalloc(&pdev->dev, struct_size(clk_data, hw_data.hws, num_clks + 2),
		GFP_KERNEL);
```

Reply:
> > Unfortunately, these defines are an ABI and therefore cannot be changed. The bug needs to be solved without changing the binding header, only by changing the driver.
> 
> Which part is dependent on this?

The bit I replied to. You cannot change the defines in include/dt-bindings/clock/microchip,mpfs-clock.h, they're part of the devicetree ABI.

Reply:
> I guess it has to allocate `num_clks + 2 (where num_clks=10)` to store `hw_data.hws[10]` and `hw_data.hws[11]` for `CLK_CCC_PLL1_OUT2/3`.
> `hw_data.hws[2]` and `hw_data.hws[3]` will be NULL.

Yah, I think so too. It's probably worth checking the behaviour if someone requests
one of the DLLs, but I think the clk core will handle there being no provider registered just fine.

Reply:
> edit: existing devicetrees rely on the defines being what they are, so they would no longer request the correct clocks if you changed the defines.

I see.
I didn't take care of devicetrees already built with older kernel, but it makes sense.

`devm_of_clk_add_hw_provider` sets `of_clk_hw_onecell_get` as getter.
`of_clk_hw_onecell_get` simplly returns hw_data->hws[idx] so it will return NULL for DLLs.
```c
struct clk_hw *
of_clk_hw_onecell_get(struct of_phandle_args *clkspec, void *data)
{
	struct clk_hw_onecell_data *hw_data = data;
	unsigned int idx = clkspec->args[0];

	if (idx >= hw_data->num) {
		pr_err("%s: invalid index %u\n", __func__, idx);
		return ERR_PTR(-EINVAL);
	}

	return hw_data->hws[idx];
}
```

`of_clk_hw_onecell_get` will be set as `get_hw` and docs states that it can return NULL.
So it should be fine to return NULL for DLLs.
`clk_hw_create_clk` checks with `IS_ERR_OR_NULL` for example.
```c
/**
 * struct of_clk_provider - Clock provider registration structure
 * @link: Entry in global list of clock providers
 * @node: Pointer to device tree node of clock provider
 * @get: Get clock callback.  Returns NULL or a struct clk for the
 *       given clock specifier
 * @get_hw: Get clk_hw callback.  Returns NULL, ERR_PTR or a
 *       struct clk_hw for the given clock specifier
 * @data: context pointer to be passed into @get callback
 */
struct of_clk_provider {
	struct list_head link;

	struct device_node *node;
	struct clk *(*get)(struct of_phandle_args *clkspec, void *data);
	struct clk_hw *(*get_hw)(struct of_phandle_args *clkspec, void *data);
	void *data;
};
```